devito/ir/support/utils.py

```diff
@@ -158,3 +159,41 @@ def detect_io(exprs, relax=False):
             writes.append(f)
 
     return filter_sorted(reads), filter_sorted(writes)
+
+
+def indexify_expr(expressions, **kwargs):
```

Comment:
Is this a good place to define this function?

Reply:
Drop "_expr" from the name

Reply:
Where should this function go is a very good question. 

`ir` uses `types`, so it can't go there, since this will have to be called from inside `ConditionalDimension`

an option would be `passes/expressions`, but then again it wouldn't be callable from within `types/dimension.py`

how about directly inside `types/equation.py` ? after all, this is to be called with eqns as input. I'm not 100% sure though...

Thoughts @mloubout @rhodrin ?

Reply:
I was thinking about that but there is another indexify fuction in manupulation.py . Other name?

Reply:
+ we already use it in ours (line:176 below)

Reply:
hmm.. if I add it in equation.py I get
```
...
  File "/home/gbisbas/workspace/devito/devito/types/equation.py", line 8, in <module>
    from devito.symbolics import retrieve_indexed, retrieve_functions, uxreplace
ImportError: cannot import name 'retrieve_indexed' from partially initialized module 'devito.symbolics' (most likely due to a circular import) (/home/gbisbas/workspace/devito/devito/symbolics/__init__.py)

```

Reply:
I thought about that and I don't have a better suggestion than putting it in `ir/equations/algorithms.py`.

Please call it `lower_exprs` or something along the lines of `lower_...`

This most likely will create a circular import when you will attempt to import it from `types/dimension.py`. I don't have a solution to this ATM. So just use a local import for now:

```
class ConditionalDimension(...):
    def __init_finalize__(...):
        ....

        from devito.ir.equatios import lower_exprs
        ...
```

@mloubout any suggestion would be appreciated, but I don't think we can escape from this, as needed it in `types`, and in particular in `types/dimension.py`, doesn't give us many alternatives I feel. Not unless we start splitting the real API type classes from the support type classes, but jeez...that's for another day/life I think

Reply:
I also think this line:

https://github.com/devitocodes/devito/pull/1276/files#diff-5ebf1ee0cae059be380b472501c3d663R275

should be squashed inside this new `lower_exprs`, at the very end:

```
...

processed = [LoweredEq(e) for e in processed if e.is_Equality]

return processed
```

Reply:
hm.. this link you pasted doens't give me a line?